Build now works on Travis Container Infrastructure

tjenkinson opened the pull request:
Tested a build locally and it still appears to work but this will need testing with different formats to make sure I haven't removed something important.

Chrome and Firefox browsers now work and are used for running the tests. I have fixed one of the tests which was failing in Firefox.

Removed jquery dependency from tests and clappr-zepto now used and required explicitly in the place it was required.

Updated some of the versions of some of the dependencies.

Removed "browser" property from `package.json`. Not sure what it was there for, or if it was important?

Moved dependencies from "dependencies" to "dev-dependencies" because I don't think any of the ones that were there were required for using Clappr, because everything that is needed is bundled into the `clappr.js` file. This means `npm-shrinkwrap.json` is empty as this by default only locks the `dependencies` list but I think this still makes sense.

tjenkinson commented:
Closes #529

leandromoreira commented:
Great work @tjenkinson =) congrats, I'll try to test this build later (with hls, mp4 ... )

tjenkinson commented:
I might be misunderstanding the difference between dev and not dev then.
My understanding is that dev dependencies are the dependencies needed for a developer to work on the project and built it etc.
`dependencies` are the dependencies that are required in order for the project to be run/used by a user. Because the entire project is bundled into the `clappr.min.js` file this is all the user needs in order to use the project. The entry point that the end user should use to run the project is what's specified by the `main` property, which is `dist/clappr.min.js`.

For example if someone else in a separate project does `npm install clappr` they will get `clappr.min.js` and that's it. If they `require("clappr")` somewhere then it is that file which will be loaded (thanks to the `main` property). They don't need anything else to use the project.

However, if the `main` property was `src/main.js` instead, then all of the `devDependencies` would be come `dependencies` as then they would be needed when clappr is built as part of the users build process when they build their project.

leandromoreira commented:
> My understanding is that dev dependencies are the dependencies needed for a developer to work on the project and built it etc. `dependencies` are the dependencies that are required in order for the project to be run/used by a user. 

I think the same.

> Because the entire project is bundled into the clappr.min.js file this is all the user needs in order to use the project. The entry point that the end user should use to run the project is what's specified by the main property, which is dist/clappr.min.js.

Yes too.

> For example if someone else in a separate project does `npm install clappr` they will get clappr.min.js and that's it.

Yes too but I think if you move those dependencies to `devDependencies`, the final bundle **clappr.min.js** will throw an error because `uniq` and the others are not contained within it. 

That's my understanding (I grasp that from using: bundler, pip, maven, gradle ...) but I really don't have 100% confidence with `npm`. What you can do to test is:

```shell
vim package.json #(move everything to devDependency)
rm -rf node_modules && rm npm-shrinkwrap.json
npm install
npm start
```
And try to see on browser if there is any error.

tjenkinson commented:
I just tried and don't get any errors. If `uniq`is `require()`'d in a file my understanding was that webpack would then bundle it into the built javascript file.

An exception being that if the webpack config declared `uniq` as being an `external`, meaning it should expect it to be provided some other way. E.g as a global variable or exported from some other javascript on the page. The place it would expect it to be found depends on some other webpack options.

leandromoreira commented:
You're right (that's crazy), as I said of all those package managers (bundler, maven, ...) all of them exclude libs not declared as `dependency`. In the end webpack will do the job of building clappr.min.js

But anyway, I still think we need to **document** the intention under `dependency` because declaring it as [`devDependency`](http://stackoverflow.com/a/19341028) we're assuming that they are not required for production.

>**devDependencies** are for the development-related scripts, e.g. unit testing, packaging scripts, documentation generation, etc.

>**dependencies** are required for production use, and assumed required for dev as well.

towerz commented:
@tjenkinson actually you are correct in your understanding. Clappr declares the main module as clappr.min.js in package.json (which makes unnecessary to declare its dependencies), but its source code is included in the npm package as well. This allows a user to include Clappr in a project codebase, generating a single bundle.

So in order to use Clappr by importing its source, it would be necessary to redeclare its own dependencies in the target project. Using `devDependencies` doesn't fit this well, because the user won't be able to differentiate what is required for building Clappr from what's only a development tool used in the Clappr project. In this case, using `dependencies` instead of `devDependencies` serves this purpose well, like @leandromoreira mentioned.

In fact, we only declare clappr.min.js as entry point in order to allow projects which do not use ES6 to include Clappr as dependency without needing to compile it.

tjenkinson commented:
@towerz so do you think `main` should point to `src/main.js` then?

HLSJS now works like this so webpack is actually building that as well as clappr when it's built.

If someone doesn't want to build clappr themselves they could then set their project configs to use the version in the dist folder?

I think `main`, `dependencies` and `devDependencies` are all related and having `main` point to a built version but still include the dependencies to build doesn't really make sense. For example if someone wants to include clappr in a project with webpack and built it themselves, with `main` pointing to the built version I think they will have to mess with their configs to get it to work and override npm's behaviour of looking at `main`.

This doesn't currently effect me because I'm just including the built version manually but if I was using npm I think this is how I'd expect it to work. Maybe I'm wrong.

towerz commented:
@tjenkinson we had already done that before, pointing main to src/main.js, but there were some complaints after we migrated to webpack. There were people who used Clappr with browserify that weren't able to build their projects without including babel and other dev dependencies, because they couldn't set their projects to use clappr.min.js easily, at least not without breaking asset references (like swf paths, etc). This was described in #533.

Since changing this back to clappr.min.js didn't affect us, we gave in and changed the main back to the build output dist/clappr.min.js. We can check whether these problems do still occur and change main back to src/main.js.

tjenkinson commented:
Looking at that I think if `browser` was set to `dist/clappr.min.js` I don't see why it wouldn't work for browserify users. Not tested it though.

They will probably have to set the `baseUrl` option to the location they've copied their assets to. Don't think browserify will copy assets so that would have to be part of the build script.

tjenkinson commented:
It would mean though that even if a user is using browserify, and therefore going to use the built version as a result of `browser`, then are still going to get the build dependencies on `npm install` even though they're never going to use them. Guess that's just a side effect of the `browser` 'hack' though.

towerz commented:
Yes, it seems to be the case. And to make things worse, with browserify the user has to manually manage the player assets when compiling the bundle, making it very difficult to avoid errors.

tjenkinson commented:
Yeh I started using browserify for something and was doing that, and then found webpack and switched because it was so much better at managing everything.

I think the priority should be making sure the combination of `main`, `devDependencies`, and `dependencies` is valid for npm, over making it work for browserify because
- if `browser` is added it should still work for browserify users, even if not ideal
- I don't think it's too complicated to switch a project from using browserify to webpack. I did it and don't remember it being that difficult, although it was a small project. (https://webpack.github.io/docs/webpack-for-browserify-users.html)
- it means people using npm will get the expected behaviour, instead of having to make tweaks in their configs to make it behave how they would have expected.

Just my opinion though.

(For point 2 this is kind of already happening with hlsjs because that uses browserify but is being built with webpack fine here).

towerz commented:
I agree it's not complicated to switch from browserify to webpack, but I don't want to force this change on anyone. About your first point, if `browser` is added, it doesn't mean that will work for browserify users (it will still present the assets issue), but we can add instructions for browserify users to do that.

If you want to proceed with that, I'd like to ask you to revert the devDependency changes and switch the main reference to src/main.js, and add instructions to make browserify work correctly with assets in the readme file.

tjenkinson commented:
Yes that's fine I'll do that.

Won't the assets issue still be the same as it is now though? I don't think the combination of these changes will make a difference.

> On 24 Nov 2015, at 18:23, Bruno Torres <notifications@github.com> wrote:
> 
> I agree it's not complicated to switch from browserify to webpack, but I don't want to force this change on anyone. About your first point, if browser is added, it doesn't mean that will work for browserify users (it will still present the assets issue), but we can add instructions for browserify users to do that.
> 
> If you want to proceed with that, I'd like to ask you to revert the devDependency changes and switch the main reference to src/main.js, and add instructions to make browserify work correctly with assets in the readme file.
> 
> —
> Reply to this email directly or view it on GitHub.
>

towerz commented:
I'm not sure, theoretically it shouldn't make a difference, but since people have had problems with this change before, perhaps we need to provide clearer instructions for using Clappr with browserify. I just asked you to do so because you are already changing the package spec, and I think we need to better document these changes during the process.

tjenkinson commented:
Just looking into this further and there it looks like webpack supports `browser` as well along with a few others, and by default webpack will also default to `browser` and use the built version, which might be what people want.
https://webpack.github.io/docs/configuration.html#resolve-packagemains

I've opened an issue in webpack (https://github.com/webpack/webpack/issues/1671) to see if it's possible at the moment to override which property to use in `package.json` for specific modules, to see if it's easy for developers to opt to build it themselves if they wanted.
https://github.com/webpack/webpack/issues/1671

towerz commented:
I wouldn't recommend to use the built version when using webpack, unless it's declared as `external`. It would introduce some unnecessary overhead to the built product.

In webpack, to override the package main when building, we use the following alias in webpack.config.js in our custom player:
```javascript
resolve: {
    alias: { Clappr: 'clappr/src/main.js' },
    root: [path.resolve(__dirname, 'node_modules/clappr/src')],
    extensions: ['', '.js'],
}
```
This makes webpack use src/main.js as Clappr package entry point (when requiring/importing 'Clappr') and use clappr/src as a resolve path (in order to resolve 'plugins/poster' for example).

tjenkinson commented:
Great ok so I guess if `main` became `./src/main.js` and `browser` `./dist/clappr.min.js` that would be what people would have to do and could go in the readme, if they want to build themselves.

Would be a nice feature in weback though if you could do something like:
```
resolve: {
    moduleName: 'Clappr',
    resolve.packageMains: ['main']
}
```

To achieve the same thing because then you don't have to rely on the source entry point always being at that location, and if it changes it wouldn't matter. That's what I'm wondering in the ticket I just opened.

tjenkinson commented:
I've moved the dependencies back and updated the readme. @towerz is there any other config you had to add or was this enough? https://github.com/clappr/clappr/blob/travis-container-infrastructure2/README.md#installing-for-webpack

leandromoreira commented on .travis.yml:
```diff
@@ -1,24 +1,13 @@
+  - "4.2.2"
 
-env:
-  - CHROME_BIN=/usr/bin/chromium-browser COVERALLS_SERVICE_NAME=travis-ci COVERALLS_REPO_TOKEN=sj8SOxUBGaVeFGSY9ryrC2eTFpVjCPHJl
```
I think it's still good to send coveralls, but we can do that after the merge :+1:

tjenkinson replied:
It's back now :) wasn't sure if it was doing anything before
